Source/MetroXHunter/Reload/ReloadComponent.cpp

```diff
@@ -0,0 +1,359 @@
+	}
+
+	GetWorld()->GetTimerManager().SetTimer(
+		TimerHandleReloadPlayback, [this]()
+		{ bIsReloadActive = false; },
+		FinalWaitingTime,
+		false
+	);
+}
+
+void UReloadComponent::UpdateCurrentGunState( EGunState NewState )
+{
+	CurrentGunState = NewState;
+}
+
+void UReloadComponent::RetrievePlayerInventory()
+{
+	AActor* Owner = GetOwner();
+	if ( Owner )
+	{
+		PlayerInventory = Cast<UInventoryComponent>( Owner->GetComponentByClass( UInventoryComponent::StaticClass() ) );
+	}
+}
+
+
+void UReloadComponent::RetrieveHUD()
+{
+	PlayerController = UGameplayStatics::GetPlayerController( this, 0 );
+	if ( PlayerController )
+	{
+		HUD = PlayerController->GetHUD();
+	}
+}
+
+void UReloadComponent::InitializeReloadData( UReloadData* NewReloadData )
+{
+	if ( NewReloadData )
+	{
+		ReloadDataAsset = NewReloadData;
+	}
+}
+
+void UReloadComponent::GetNormalizedReloadTimings( 
+	float& OutPerfectReloadStartTime, 
+	float& OutActiveReloadStartTime, 
+	float& OutActiveReloadEndTime 
+) const
+{
+	if ( ReloadDataAsset && ReloadDataAsset->NormalReloadDuration > 0.0f )
+	{
+		OutPerfectReloadStartTime = ReloadDataAsset->PerfectReloadStartTime / ReloadDataAsset->NormalReloadDuration;
+		OutActiveReloadStartTime = ReloadDataAsset->ActiveReloadStartTime / ReloadDataAsset->NormalReloadDuration;
+		OutActiveReloadEndTime = ReloadDataAsset->ActiveReloadEndTime / ReloadDataAsset->NormalReloadDuration;
+	}
+	else
+	{
+		OutPerfectReloadStartTime = 0.0f;
+		OutActiveReloadStartTime = 0.0f;
+		OutActiveReloadEndTime = 0.0f;
+	}
+}
+
+float UReloadComponent::GetNormalizedReloadElapsedTime() const
+{
+	if ( ReloadDataAsset && ReloadDataAsset->NormalReloadDuration > 0.0f )
+	{
+		return ReloadElapsedTime / ReloadDataAsset->NormalReloadDuration;
+	}
+	else
+	{
+		return 0.0f;
+	}
+}
+
+void UReloadComponent::UpdateAmmoCount( int NewCount )
+{
+	CurrentAmmoInMagazine = NewCount;
+
+	OnAmmoCountUpdated.Broadcast();
+}
+
+void UReloadComponent::GetAmmoData( int& CurrentAmmo, int& MaxAmmo ) const
+{
+	CurrentAmmo = CurrentAmmoInMagazine;
+	MaxAmmo = MaxAmmoInMagazine;
+}
+
+// Computes the new ammo count after reloading based on available inventory and magazine space
+void UReloadComponent::ComputeReloadAmmoCount( int& NewMagazineAmmoCount, int& InventoryAmmoConsumed )
+{
+	int AmmoToConsumeToMax = MaxAmmoInMagazine - CurrentAmmoInMagazine;
+
+	// Obtenir les munitions disponibles dans l'inventaire
+	int InventoryAmmo = PlayerInventory ? PlayerInventory->GetCurrentAmmoAmount() : 0;
+
+	if ( InventoryAmmo >= AmmoToConsumeToMax )
+	{
+		// Recharge compl�te possible
+		NewMagazineAmmoCount = MaxAmmoInMagazine;
+		InventoryAmmoConsumed = AmmoToConsumeToMax;
+	}
+	else
+	{
+		// Recharge partielle
+		NewMagazineAmmoCount = CurrentAmmoInMagazine + InventoryAmmo;
+		InventoryAmmoConsumed = InventoryAmmo;
+	}
+}
+
+
+void UReloadComponent::DecrementAmmo()
+{
+	if ( CurrentAmmoInMagazine > 0 )
+	{
+		--CurrentAmmoInMagazine;
+		OnAmmoCountUpdated.Broadcast();
```

Comment:
I think it would be better to funnel the code by using `UpdateAmmoCount( CurrentAmmoInMagazine - 1 )` instead; so we don't have multiple code that individually change the same variable and call the event